Combine H1 and H2 conn_pool implementations

ggreenway opened the pull request:
This allows the H2 connection pool to have more than 2 connections (1 active, 1 draining), which allows things like very low values of `max_requests_per_connection` to work because there can be multiple draining connections concurrently, as well as supporting multiple active connections.

Risk Level: High
Testing: Unit, integration
Docs Changes: Added to .proto; updated conn pool and circuit breaker docs; added old behavior to deprecation list
Release Notes: added
Fixes #7403
Fixes #9215

ggreenway commented:
cc @snowp

ggreenway commented:
I have 2 big concerns with this PR:

1) I messed up the implementation and it completely breaks the world until the bug is fixed
2) Subtle behavior changes to existing configurations. Behavior is mostly the same for both H1 and H2 pools, with the exception of circuit breakers. The old behavior was to only look at connection circuit breakers for H1 pools, and only look at request circuit breakers for H2 pools. Now, both types are looked at in all cases. Will this break someone?

snowp commented:
Re 2. I have written code myself that doesn't touch the unused circuit breaker for H1/H2 only clusters, so I'm sure there are several cases where people have bumped the max connection CB for a HTTP/1.1 only cluster to permit some concurrency but left the max request CB limit at the default. I'm wondering if we should be runtime guarding the new behavior and then communicate the change of behavior with a timeline for changing the default. Thoughts @mattklein123 ?

snowp commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+      Envoy::Upstream::reportUpstreamCxDestroyActiveRequest(host_, event);
+    }
+
+    const bool check_for_drained = client.closingWithIncompleteRequest();
```
you're calling `closingWithIncompleteRequest` in the previous if block as well, should probably consolidate?

snowp commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+    // We need to release our resourceManager() resources before checking below for
+    // whether we can create a new connection. Normally this would happen when
+    // client's destructor runs, but this object needs to be deferredDelete'd(), so
+    // this forces part of it's cleanup to happen now.
```
`it's` -> `its`

snowp commented on test/test_common/utility.cc:
```diff
@@ -311,16 +311,22 @@ std::string TestUtility::convertTime(const std::string& input, const std::string
 // static
-bool TestUtility::gaugesZeroed(const std::vector<Stats::GaugeSharedPtr>& gauges) {
-  // Returns true if all gauges are 0 except the circuit_breaker remaining resource
+std::string TestUtility::nonZeroedGauges(const std::vector<Stats::GaugeSharedPtr>& gauges) {
```
seems like the only use of this function can be better expressed with the old function?

ggreenway replied:
This version of the function prints out which stat was non-zero and what it's value was when the test expectation fails (this function doesn't actually print anything, but the test expectation itself is `EXPECT_EQ(empty_string, nonZeroedGauges())`, which prints the string when it isn't empty). With the old version, I had to attach a debugger to get that information, so I found this version much easier to iterate on test failures with.

ggreenway replied:
There may be a better gtest/gmock way to do this, and I played with returning a vector of something that the expectation could print out, but it was more work than it was worth (to me) to figure out how to make it print nicely (not print pointers; print names of stats), so I just did this slightly clunky version with a string.

ggreenway commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+void ConnPoolImplBase::createNewConnection() {
+  ENVOY_LOG(debug, "creating a new connection");
+  ActiveClientPtr client = instantiateActiveClient();
+  client->state_ = ActiveClient::State::CONNECTING;
```
That is the default value; I'll change it to an ASSERT.

ggreenway commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+  auto& old_list = owningList(client.state_);
+  auto& new_list = owningList(new_state);
+  client.state_ = new_state;
+  if (&old_list != &new_list) {
```
Yes, when transitioning from busy to draining.

I looked at the c++ sodlib docs, and I couldn't tell if it was allowed for these two to be equal, so I added the check to make sure it was safe.

ggreenway commented:
> Might be good to add some test coverage to the new circuit breaking behavior for H1/H2

I wanted to discuss and figure out what we want the behavior to be before I write tests for it.

alyssawilk commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+      host_->cluster().stats().upstream_cx_overflow_.inc();
+    }
+
+    // If we have no connections at all, make one no matter what so we don't starve.
```
It's a bit non-obvious why we always create connections but we don't always create requests in the else block before.  mind adding a bit more detail to the comments?

mattklein123 replied:
I know where this code came from before, but I'm wondering with this new code if this is still required? Meaning, how is it possible that we would have zero connections not be able to make a connection? That seems like a config error that should be caught elsewhere. I think we can probably get rid of this now?

ggreenway replied:
I don't know what this was for originally, but one case that comes to mind is if we hit the connection circuit breaker due to many active/draining connections for other pool members, and are unable to create a single connection to this pool member. I'm not sure if that case requires any special case or not. I can delete this if you're sure it's not needed.

mattklein123 replied:
> but one case that comes to mind is if we hit the connection circuit breaker due to many active/draining connections for other pool members, and are unable to create a single connection to this pool member.

Hmm right, I think that might still be needed. Can you add more comments here and we can decide later whether we still need this or not?

alyssawilk commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+    return true;
+  }
+
+  for (const auto& ready_client : ready_clients_) {
```
Do we think this scales sufficiently?  It looks like this is called by freeOnePool which is walking pools looking for something unused, and given I've seen use cases of hundreds to thousands of endpoints and tens of thousands of connections I'm mildly wary of looping if it'll be expensive on busy machines.

mattklein123 replied:
Yeah good call out. I think we could keep this state actively? At minimum can we TODO this?

ggreenway replied:
Good catch.

I think in the common case this shouldn't be too expensive, because it'll stop iterating as soon as it finds a connection with an active request. But there could be cases where this is too expensive.

alyssawilk commented on source/common/http/conn_pool_base.cc:
```diff
@@ -89,5 +341,65 @@ void ConnPoolImplBase::onPendingRequestCancel(PendingRequest& request) {
+      parent_.host_->cluster().stats().upstream_cx_length_ms_, parent_.dispatcher_.timeSource());
+  connect_timer_->enableTimer(parent_.host_->cluster().connectTimeout());
+
+  parent_.host_->stats().cx_total_.inc();
```
Lazy check - are all these stats tested in existing H2 tests?

ggreenway replied:
Missing test coverage:
* upstream_cx_active_

I'll add a case for it

alyssawilk commented on source/common/http/conn_pool_base.h:
```diff
@@ -64,15 +115,43 @@ class ConnPoolImplBase : protected Logger::Loggable<Logger::Id::pool> {
+  // Closes any idle connections.
+  void closeIdleConnections();
 
+  // Called by derived classes anytime a request is completed or destroyed for any reason.
```
any time

mattklein123 commented on api/envoy/config/core/v3alpha/protocol.proto:
```diff
@@ -136,6 +136,10 @@ message Http2ProtocolOptions {
+  //
+  // For upstream connections, this also limits how many streams Envoy will initiate concurrently
+  // on a single connection. If the limit is reached, Envoy may queue requests or establish
+  // additional connections (as allowed per circuit breaker limits).
```
> Given this is really big refactor, I'd think we want to at least load test, at best smoke test in prod before merging.
@htuch and @mattklein123 what do you think of asking the folks who are willing to test security patches if they're willing (optionally) to try this out as well?

This is scary enough that I would actually be inclined to potentially copy the code for the conn pool and have a runtime feature kill switch to go back to the old code. If we think that would be more effort/muck than it's worth than yeah we definitely need to get some smoke testing on this for sure.

ggreenway replied:
I think a complete copy of the old version with a runtime switch is a reasonable way to do this. Then delete the old version according to the deprecation policy in a few versions. 

Would the kill switch be a CLI arg, or configuration?

Managing both versions of the tests may be messy, but manageable.

mattklein123 replied:
> Would the kill switch be a CLI arg, or configuration?

I would just do it with the new `runtimeFeatureEnabled` support that @alyssawilk added, defaulting to on.

ggreenway replied:
I'll have to read about this new awesomeness

ggreenway replied:
Ok, I pushed a commit that puts all of the old code back with a runtimeFeatureEnabled to switch between them. Is this what you were hoping for?

mattklein123 commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+
+void ConnPoolImplBase::closeIdleConnections() {
+  // Create a separate list of elements to close to avoid mutate-while-iterating problems.
+  std::list<ActiveClient*> toClose;
```
nit: to_close

ggreenway commented on api/envoy/api/v2/core/protocol.proto:
```diff
@@ -117,6 +117,10 @@ message Http2ProtocolOptions {
+  //
+  // For upstream connections, this also limits how many streams Envoy will initiate concurrently
+  // on a single connection. If the limit is reached, Envoy may queue requests or establish
+  // additional connections (as allowed per circuit breaker limits).
```
This is in message `Http2ProtocolOptions`, so not applicable to H1.

ggreenway commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,292 @@
+    transitionActiveClientState(client, ActiveClient::State::READY);
+    if (!delay_attaching_request) {
+      onUpstreamReady();
+    }
```
H2 always passes `false` for `delay_attaching_request`. 

See comment from http1/conn_pool.cc:
```
ConnPoolImpl::StreamWrapper::~StreamWrapper() {
  // Upstream connection might be closed right after response is complete. Setting delay=true
  // here to attach pending requests in next dispatcher loop to handle that case.
  // https://github.com/envoyproxy/envoy/issues/2715
  parent_.parent().onRequestClosed(parent_, true);
}
```

It's tempting to copy this comment to here also. Thoughts?

ggreenway replied:
There's several other valid states that don't require any additional action:
* client.state_ == READY (this means the connection was already ready to receive additional requests; it's current request count is >=0 and < max_concurrent_requests)
* client.state_ == DRAINING && numActiveRequests > 0 (still needs more draining)
* Possibly others, but those two came to mind quickly

mattklein123 commented on api/envoy/api/v2/core/protocol.proto:
```diff
@@ -117,6 +117,10 @@ message Http2ProtocolOptions {
   // allowed for peer on one HTTP/2 connection. Valid values range from 1 to 2147483647 (2^31 - 1)
   // and defaults to 2147483647.
+  //
+  // For upstream connections, this also limits how many streams Envoy will initiate concurrently
```
Can you add a version history note about this change?

ggreenway replied:
As in, "Beginning with version 1.14: For upstream connections, this also......."?

mattklein123 commented on source/common/http/conn_pool_base.cc:
```diff
@@ -1,40 +1,293 @@
+        host_->cluster().resourceManager(priority_).requests().canCreate() &&
+        host_->cluster().resourceManager(priority_).connections().canCreate()) {
+      createNewConnection();
+    }
```
Coming back to this discussion, if the circuit breaker check here fails, what happens? I think we could still starve? (This is the block that used to create a connection no matter what, right? Or is that somewhere else)?

ggreenway replied:
No, the block that unconditionally makes the connection didn't move; it was always in `newStream()`. In the original http1 pool code, it has this check for this case: `if (pending_requests_.size() > (ready_clients_.size() + busy_clients_.size())) {`. I think the presumption is that if we just closed a connection, there should be available circuit breaker capacity for 1 more connection.

In the new version, I think it would only be starved if there are no connections in this pool, so I think it needs roughly the same check as in `newStream()`. I'll sort that out.

ggreenway replied:
Also, while thinking this through, I don't think the requests circuit breaker should be checked here at all. The behavior has always been that if a connection is ready, then requests are attached to it, and at that time if the requests circuit breaker is exceeded, an `onPoolFailure(Overflow)` is issued. That code happens in `attachRequestToClient()`, so I think that's the only place we should check that circuit breaker.

mattklein123 commented on source/common/http/conn_pool_base.h:
```diff
@@ -1,43 +1,85 @@
+    virtual bool closingWithIncompleteRequest() const PURE;
+    virtual StreamEncoder& newStreamEncoder(StreamDecoder& response_decoder) PURE;
+
+    enum class State { CONNECTING, READY, BUSY, DRAINING, CLOSED };
```
nit: might be nice to have some comments on what each of these states mean

mattklein123 commented on source/common/runtime/runtime_features.cc:
```diff
@@ -32,6 +32,8 @@ constexpr const char* runtime_features[] = {
     "envoy.reloadable_features.reject_unsupported_transfer_encodings",
     "envoy.reloadable_features.strict_method_validation",
+    "envoy.reloadable_features.new_http1_connection_pool_behavior",
+    "envoy.reloadable_features.new_http2_connection_pool_behavior",
```
IMO it's fine to land as default on, though I think we might want to email envoy-dev and/or ask both Pinterest and DropBox to smoke test for us.

ggreenway commented on source/common/runtime/runtime_features.cc:
```diff
@@ -32,6 +32,8 @@ constexpr const char* runtime_features[] = {
     "envoy.reloadable_features.reject_unsupported_transfer_encodings",
     "envoy.reloadable_features.strict_method_validation",
+    "envoy.reloadable_features.new_http1_connection_pool_behavior",
+    "envoy.reloadable_features.new_http2_connection_pool_behavior",
```
I need to revise the existing connection pools docs also

snowp commented on source/common/http/http1/conn_pool.h:
```diff
@@ -83,50 +74,27 @@ class ConnPoolImpl : public ConnectionPool::Instance, public ConnPoolImplBase {
+  struct ActiveClient : public ConnPoolImplBase::ActiveClient {
     ActiveClient(ConnPoolImpl& parent);
-    ~ActiveClient() override;
+    ~ActiveClient() override = default;
```
i dont think this is needed anymore

snowp commented on test/common/http/http2/conn_pool_test.cc:
```diff
@@ -218,24 +272,78 @@ TEST_F(Http2ConnPoolImplTest, DrainConnections) {
+  InSequence s;
+
+  {
+    // Create request and complete it
```
nit: end comments with periods in this file

ggreenway commented on source/common/http/http1/conn_pool.cc:
```diff
@@ -351,6 +142,21 @@ CodecClientPtr ProdConnPoolImpl::createCodecClient(Upstream::Host::CreateConnect
+                 Upstream::ResourcePriority priority,
+                 const Network::ConnectionSocket::OptionsSharedPtr& options,
+                 const Network::TransportSocketOptionsSharedPtr& transport_socket_options) {
+  if (Runtime::runtimeFeatureEnabled(
```
Yeah, this is in-between some of the flags which toggle behavior immediately, and ones that require an envoy restart. @alyssawilk any thought on what to say about this and where?

mattklein123 replied:
I would just mention this in the release note via my other comment on that file.

ggreenway commented:
I found (and fixed) a bug that wasn't seen because the default resourceManager in tests only allows 1 connection. I increased that for h2 tests and fixed the bug where it was creating too many connections when none were yet connected, and added more tests around that case.

alyssawilk commented:
CI looks unhappy - real or flake?

ggreenway commented:
It's in the h1 conn pool test, so probably not a flake. But it passed on my machine. I'm digging into it.

ggreenway commented:
/waiting

ggreenway commented:
/wait

ggreenway commented:
> But it passed on my machine.

When I ran the tests against the wrong branch. Oops.

When there is a pending request, and a connection in the process of connecting, if `drainConnection()` or `addDrainedCallback()` is called, should the pending connection be closed, or kept open until it is connected to try to serve the pending request? @mattklein123

mattklein123 commented:
> When there is a pending request, and a connection in the process of connecting, if drainConnection() or addDrainedCallback() is called, should the pending connection be closed, or kept open until it is connected to try to serve the pending request? @mattklein123

As long as there are pending requests I think we have to allow them to be served, which includes waiting for connection?

ggreenway commented:
That makes sense. I hadn't realized that the two functions serve very different use cases. I think I have the desired behavior for both now.

ggreenway commented:
@envoyproxy/api-shepherds PTAL at this. It's not a change to the protos, just a doc change and applying an existing field in a new context.

mattklein123 commented on docs/root/intro/arch_overview/upstream/circuit_breaking.rst:
```diff
@@ -3,6 +3,14 @@
+  is needed, disable runtime feature "envoy.reloadable_features.new_http1_connection_pool_behavior" or
+  "envoy.reloadable_features.new_http2_connection_pool_behavior" and then re-configure your clusters or
+  restart Envoy. Note that the old behavior is deprecated and will be removed according to the deprecation
+  policy.
```
My feeling is to just kill this attention block. The new behavior is documented below and we can document the changes in the release notes? WDYT?

ggreenway replied:
I think that's fine.  A new user will come to this page, read the docs, and set the correct circuit breaker settings. An upgrader whose config no longer works will probably check release notes first.

mattklein123 commented on docs/root/intro/deprecated.rst:
```diff
@@ -12,6 +12,10 @@ Deprecated items below are listed in chronological order.
+* The previous behavior for upstream connection pool circuit breaking described
+  `here <https://www.envoyproxy.io/docs/envoy/v1.13.0/intro/arch_overview/upstream/circuit_breaking>` has
+  been deprecated in favor of the new behavior described
+  `here <https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking>`.
```
nit: switch this one to a direct ref link.

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -5,6 +5,7 @@ Version history
 ================
 * config: use type URL to select an extension whenever the config type URL (or its previous versions) uniquely identify a typed extension, see :ref:`extension configuration <config_overview_extension_configuration>`.
 * retry: added a retry predicate that :ref:`rejects hosts based on metadata. <envoy_api_field_route.RetryPolicy.retry_host_predicate>`
+* upstream: combined http/1 and http/2 connection pool code. This means that circuit breaker limits for both requests and connections apply to both pool types. Also, http/2 now has the option to limit concurrent requests on a connection, and allow multiple draining connections.
```
This is where I would document the kill switches.
